flink-ml-lib/src/main/java/org/apache/flink/ml/recommendation/swing/Swing.java

```diff
@@ -0,0 +1,430 @@
+                                })
+                        .returns(Types.TUPLE(Types.LONG, Types.LONG));
+
+        SingleOutputStreamOperator<Tuple3<Long, Long, Map<Long, String>>> userAllItemsStream =
+                itemUsers
+                        .keyBy(tuple -> tuple.f0)
+                        .transform(
+                                "fillUserItemsTable",
+                                Types.TUPLE(
+                                        Types.LONG,
+                                        Types.LONG,
+                                        Types.MAP(Types.LONG, Types.STRING)),
+                                new CollectingUserBehavior(
+                                        getMinUserBehavior(), getMaxUserBehavior()));
+
+        RowTypeInfo outputTypeInfo =
+                new RowTypeInfo(
+                        new TypeInformation[] {
+                            BasicTypeInfo.LONG_TYPE_INFO, BasicTypeInfo.STRING_TYPE_INFO
+                        },
+                        new String[] {getItemCol(), getOutputCol()});
+
+        DataStream<Row> output =
+                userAllItemsStream
+                        .keyBy(tuple -> tuple.f1)
+                        .transform(
+                                "computingSimilarItems",
+                                outputTypeInfo,
+                                new ComputingSimilarItems(
+                                        getK(),
+                                        getMaxUserNumPerItem(),
+                                        getAlpha1(),
+                                        getAlpha2(),
+                                        getBeta()));
+
+        return new Table[] {tEnv.fromDataStream(output)};
+    }
+
+    @Override
+    public Map<Param<?>, Object> getParamMap() {
+        return paramMap;
+    }
+
+    @Override
+    public void save(String path) throws IOException {
+        ReadWriteUtils.saveMetadata(this, path);
+    }
+
+    public static Swing load(StreamTableEnvironment tEnv, String path) throws IOException {
+        return ReadWriteUtils.loadStageParam(path);
+    }
+
+    /**
+     * Appends one column, that records all items the user has purchased, to the input table.
+     *
+     * <p>During the process, this operator collect users and all items a user has purchased into a
+     * map of list. When the input is finished, this operator appends the certain
+     * user-purchased-items list to each row.
+     */
+    private static class CollectingUserBehavior
+            extends AbstractStreamOperator<Tuple3<Long, Long, Map<Long, String>>>
+            implements OneInputStreamOperator<
+                            Tuple2<Long, Long>, Tuple3<Long, Long, Map<Long, String>>>,
+                    BoundedOneInput {
+        private final int minUserItemInteraction;
+        private final int maxUserItemInteraction;
+
+        private Map<Long, Map<Long, String>> userItemsMap = new HashMap<>();
```

Comment:
The reason to use `Map` instead of `Set` is to update`userAllItemsMapState` by value of `userItemsMap`. And type of `userAllItemsMapState` cannot be `ListState<Map<Long, Set<Long>>>` because Flink doesn't have TypeInformation of `Set` type.

Reply:
Thanks for your explanation. Maybe we can add some explanation in the comments here, because it is a little confusing. 

nits: BTW, Flink can support such usage by `MULTISET`. Although it is not better than `Map<Long, String>`, but it comes from official APIs.